Detectors/ITSMFT/ITS/reconstruction/src/CA/ClusterLines.cxx

```diff
@@ -62,54 +62,54 @@ bool Line::areParallel(const Line& firstLine, const Line& secondLine, const floa
-    } else {
-      return -1;
-    }
+  std::array<float, 3> normalVector;
+  normalVector[0] = firstLine.cosinesDirector[1] * secondLine.cosinesDirector[2] -
+                    firstLine.cosinesDirector[2] * secondLine.cosinesDirector[1];
+  normalVector[1] = -firstLine.cosinesDirector[0] * secondLine.cosinesDirector[2] +
+                    firstLine.cosinesDirector[2] * secondLine.cosinesDirector[0];
+  normalVector[2] = firstLine.cosinesDirector[0] * secondLine.cosinesDirector[1] -
+                    firstLine.cosinesDirector[1] * secondLine.cosinesDirector[0];
+
+  float norm{ 0.f }, distance{ 0.f };
+  for (int i{ 0 }; i < 3; ++i) {
+    norm += normalVector[i] * normalVector[i];
+    distance += (secondLine.originPoint[i] - firstLine.originPoint[i]) * normalVector[i];
+  }
+  if (norm > precision) {
+    return std::abs(distance / std::sqrt(norm));
   } else {
-    std::array<float, 3> perpendicularVector{};
-    perpendicularVector[0] = firstLine.cosinesDirector[1] * secondLine.cosinesDirector[2] -
-                             firstLine.cosinesDirector[2] * secondLine.cosinesDirector[1];
-    perpendicularVector[1] = -firstLine.cosinesDirector[0] * secondLine.cosinesDirector[2] +
-                             firstLine.cosinesDirector[2] * secondLine.cosinesDirector[0];
-    perpendicularVector[2] = firstLine.cosinesDirector[0] * secondLine.cosinesDirector[1] -
-                             firstLine.cosinesDirector[1] * secondLine.cosinesDirector[0];
-
-    float normFirstVector{ 0 }, distance{ 0 };
-    for (int i{ 0 }; i < 3; ++i) {
-      normFirstVector += perpendicularVector[i] * perpendicularVector[i];
-      distance += (firstLine.originPoint[i] - secondLine.originPoint[i]) * perpendicularVector[i];
-    }
-    if (std::abs(normFirstVector) > precision) {
-      return std::abs(distance / std::sqrt(normFirstVector));
-    } else
-      return -1;
+    return getDistanceFromPoint(firstLine, secondLine.originPoint);
   }
 }
 
 float Line::getDistanceFromPoint(const Line& line, const std::array<float, 3> point)
 {
-  Line tmpLine{};
-  tmpLine.originPoint = point;
-  for (int index{ 0 }; index < 3; ++index)
-    tmpLine.cosinesDirector[index] = line.cosinesDirector[index];
-  float inverseNorm{ 1.f / std::sqrt(tmpLine.cosinesDirector[0] * tmpLine.cosinesDirector[0] +
-                                     tmpLine.cosinesDirector[1] * tmpLine.cosinesDirector[1] +
-                                     tmpLine.cosinesDirector[2] * tmpLine.cosinesDirector[2]) };
-  for (int index{ 0 }; index < 3; ++index)
-    tmpLine.cosinesDirector[index] *= inverseNorm;
-  return getDCA(line, tmpLine);
+  float DCASquared{ 0 };
+  float cdelta{ 0 };
+  for (int i{ 0 }; i < 3; ++i)
+    cdelta -= line.cosinesDirector[i] * (line.originPoint[i] - point[i]);
+  for (int i{ 0 }; i < 3; ++i) {
+    DCASquared += (line.originPoint[i] - point[i] + line.cosinesDirector[i] * cdelta) *
```

Comment:
can you calculate ````(line.originPoint[i] - point[i] + line.cosinesDirector[i] * cdelta)```` and then square it?

Similar comments to
L39 Line::areParallel, the same multiplication is done twice

L125 ClusterLines::ClusterLines
The covarianceFirst  and covarianceSecond are all = 1, why do you need to multiply by them?
In any case, I see that lot of multiplications can be spared by precalculating a few combinations in advance.

Idem for
L208  ClusterLines::add
L255 ClusterLines::computeClusterCentroid()

Reply:
Hi @shahor02 ,
- *"The covarianceFirst and covarianceSecond are all = 1, why do you need to multiply by them?'*

For the moment the covariance is not calculated, but I wanted to include that dummy matrix in order to implement the whole mathematics that eventually the vertexer has to do in its final state. Hence, this is done on purpose, in order to cope with what I think it is the most computing-expensive scenario.
Thus, In case at some point I would drop all this calculations, I am pretty confident that performances can only increase.

- *can you calculate (line.originPoint[i] - point[i] + line.cosinesDirector[i] * cdelta) and then square it?*

It looks like the compiler treats these two implementations in the same fashion: it always produce the same assembly. 
An example can be found [here](https://godbolt.org/g/acbiHc).

Cheers,
Matteo

Reply:
@mconcas indeed, in this case was smart enough, but it is not for the factors scattered at different lines: https://godbolt.org/g/acbiHc
Anyway, I agree it is not a blocker for the PR, but eventually such pieces should be optimized